### PR TITLE
[FSDP2] Added reduce-scatter

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118584
* #118298
* #118755
* #118223
* #118136
* #118118
* #118017
* #118004
* __->__ #117975
* #117973
* #117955
* #117950

This PR adds the FSDP reduce-scatter (the copy-in/reduce-scatter collective/view-out).
- We use gradient pre- and post-divide factors like existing FSDP (mainly for fp16 reduction).
- We use a separate CUDA stream for the reduce-scatter to conveniently handle additional kernels surrounding the collective as a separate 'thread of execution' (e.g. pre/post-divide and later the D2H gradient offload).
- ~~The implementation in this PR is more complicated to _try_ to reduce CPU overhead by using `torch.split` instead of a Python for-loop. The challenge comes from the fact that the autograd-computed unsharded gradients do not have padding. We prefer to not do an intermediate padding step and instead directly copying to the big reduce-scatter input.~~ For simplicity, I changed the implementation to include intermediate padding steps, as it can still achieve ~250 GB/s, and it avoids any `O(NP)` tensor materialization for world size `N` and `P` `nn.Parameter`s.

<details>
<summary> Recall: Copy-in/All-Gather/Copy-Out Example </summary>

Suppose we have 2 parameters with shapes `(3, 3)` (denoted with `A`s) and `(2, 2)` (denoted with `B`s) and 2 ranks, where `P` represents padding and `E` represents empty:
```
Given:
(3, 3): AAAAAAAAA
(2, 2): BBBB

Sharded parameters/all-gather inputs:
Rank 0: AAAAAA, BB
Rank 1: AAAPPP, BB

Each rank allocate group's all-gather output:
EEEEEEEEEEEEEEEE
Each rank copy-in:
Rank 0: AAAAAABBEEEEEEEE
Rank 1: EEEEEEEEAAAPPPBB

Each rank all-gather:
Rank 0: AAAAAABBAAAPPPBB
Rank 1: AAAAAABBAAAPPPBB

Each rank copy-out:
Rank 0: AAAAAAAAAPPP, BBBB
Rank 1: AAAAAAAAAPPP, BBBB
```
</details>

<details>
<summary> Copy-in/Reduce-Scatter/View-Out Example </summary>

Suppose we have 2 gradients with shapes `(3, 3)` (denoted with `a`s when not-yet-reduced and `A`s after reduced) and `(2, 2)` (denoted with `b`s and `B`s similarly) and 2 ranks, where `E` represents empty:
```
Given from autograd:
(3, 3): aaaaaaaaa
(2, 2): bbbb

Unsharded gradients/reduce-scatter inputs (no padding!):
Rank 0: aaaaaaaaa, bbbb
Rank 1: aaaaaaaaa, bbbb

Each rank allocate group's reduce-scatter input:
EEEEEEEEEEEEEEEE
Each rank copy-in:
Rank 0: aaaaaabbaaaEEEbb
Rank 1: aaaaaabbaaaEEEbb

Each rank :
Rank 0: AAAAAABBAAAEEEBB
Rank 1: AAAAAABBAAAEEEBB

Each rank view-out:
Rank 0: AAAAAA BB
Rank 1: AAA, BB
```
</details>

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 1]
```diff
@@ -105,6 +110,129 @@ def foreach_all_gather_copy_out(
...
+    device: torch.device,
+    predivide_factor: float,
+    postdivide_factor: float,
+) -> torch.cuda.Event:
+    """
+    Args:
+        unsharded_grads (List[torch.Tensor]): This list owns the references to
+            the unsharded gradients computed by autograd, so clearing this list
+            frees the gradients.
+    """
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:130]: maybe copy ``Copy-in/Reduce-Scatter/View-Out Example`` to ``foreach_reduce_scatter`` and ``foreach_reduce_scatter_copy_in`` ?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 2]
```diff
@@ -105,6 +110,129 @@ def foreach_all_gather_copy_out(
...
+        # - View out and accumulate
+        flat_grad_offset = 0  # [0, reduce_scatter_output_numel - 1]
+        for padded_unsharded_size, fsdp_param in zip(
+            padded_unsharded_sizes, fsdp_params
+        ):
+            padded_sharded_numel = padded_unsharded_size.numel() // world_size
+            sharded_numel = fsdp_param.sharded_size.numel()
+            new_sharded_grad = reduce_scatter_output[
+                flat_grad_offset : flat_grad_offset + sharded_numel
+            ].view(fsdp_param.sharded_size)
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:176]: learning purpose: I remember ``.view`` only works with contiguous memory. maybe explain more about why the memory is contiguous in ``Copy-in/Reduce-Scatter/View-Out Example`` ?

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:176]: We allocate `reduce_scatter_output` via `torch.empty()`, which will allocate it as contiguous. Viewing a 1D slice from a contiguous tensor is itself contiguous.

**Review by weifengpy — APPROVED** [review]

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 3]
```diff
@@ -105,6 +110,116 @@ def foreach_all_gather_copy_out(
...
+            (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
+        )
+        foreach_reduce_scatter_copy_in(
+            fsdp_params, unsharded_grads, reduce_scatter_input, world_size
+        )
+        _div_if_needed(reduce_scatter_input, predivide_factor)
+        # Record to mark the end of the reduce-scatter copy-in in the RS stream
+        copy_in_event = torch.cuda.Event()
+        copy_in_event.record()
+        reduce_scatter_output = torch.empty(
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:157]: nit: I changed this to `reduce_scatter_input.new_empty((reduce_scatter_output_numel,))`.

File: torch/distributed/_composable/fsdp/_fsdp_common.py [hunk 4]
```diff
@@ -86,16 +86,17 @@ def _is_composable_with_fsdp(module: nn.Module) -> bool:
 
 
 def _get_dim0_padded_size(tensor_size: torch.Size, dim0_factor: int) -> torch.Size:
```
Comment (yifuwang) [inline torch/distributed/_composable/fsdp/_fsdp_common.py:88]: nit: this gave me the impression that we have different padding rules based on the dim_0 size, which doesn't seem to be the case. Maybe consider something like this:

```python
def _get_dim0_padded_size(tensor_size: torch.Size, dim0_factor: int) -> torch.Size:
    padded_dim_0 = math.ceil(tensor_size[0] / dim_0_factor) * dim_0_factor
    return torch.Size([padded_dim_0]) + tensor_size[1:]
```

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_common.py:88]: That is clean! Do you think there is any value in fast-pathing the divisible case? There is not really any expensive CPU ops here, so it is probably fine to just keep it in one path.

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 5]
```diff
@@ -105,6 +110,116 @@ def foreach_all_gather_copy_out(
...
+    current_stream = torch.cuda.current_stream()
+    reduce_scatter_stream.wait_stream(current_stream)
+    with torch.cuda.stream(reduce_scatter_stream):
+        reduce_scatter_input = torch.empty(
+            (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
+        )
+        foreach_reduce_scatter_copy_in(
+            fsdp_params, unsharded_grads, reduce_scatter_input, world_size
+        )
+        _div_if_needed(reduce_scatter_input, predivide_factor)
```
Comment (yifuwang) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:153]: Depending on whether we end up generating triton kernels for reduce-scatter-copy-in, we might want to move this to a different place.

In case we won't be generating a triton kernel for reduce-scatter-copy-in (e.g. due to triton's limitation in handling the workload), we might want to perform the division prior to the copy-in. This way it might be picked up as an epilogue for gradient function (we can't fuse two kernels if there's an extern kernel in-between). Or we could fuse the division with reduce-scatter-copy-in.

If we can generate an efficient triton kernel for reduce-scatter-copy-in, the order won't matter.

This is not relevant to the PR. But it would be nice to note down so we know where we can hunt for perf later.

**Review by yifuwang — APPROVED** [review]
Look great!

File: torch/distributed/_composable/fsdp/_fsdp_common.py (hunk 4 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_common.py:88]: I did some microbenchmark:
- Fast path the divisible case can be ~2x faster.
- The slow path non-divisible case is faster with your code.

```
# Divisible
Yours: 0.006244088988751173 s
Original: 0.003024221397936344 s

# Not divisible
Yours: 0.00518408278003335 s
Original: 0.0058132619597017765 s
```
Script: P1148337880

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py (hunk 5 above)
Comment (yifuwang) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:153]: (This is probably pre-mature optimization thinking so don't worry too much about it)

File: torch/distributed/_composable/fsdp/_fsdp_common.py (hunk 4 above)
Comment (yifuwang) [reply torch/distributed/_composable/fsdp/_fsdp_common.py:88]: Personally I would prefer unbranched code unless there's meaningful perf implication, which I don't think exist here. But I also don't have too strong of an opinion.

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 6]
```diff
@@ -167,17 +167,19 @@ def foreach_reduce_scatter(
...
         ):
-            padded_sharded_numel = padded_unsharded_size.numel() // world_size
-            sharded_numel = fsdp_param.sharded_size.numel()
-            new_sharded_grad = reduce_scatter_output[
-                flat_grad_offset : flat_grad_offset + sharded_numel
-            ].view(fsdp_param.sharded_size)
+            new_sharded_grad = torch.as_strided(
+                reduce_scatter_output,
+                size=fsdp_param.sharded_size,
+                stride=fsdp_param.sharded_stride,
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:173]: This is very bizarre. I isolated that this change from
```
sharded_numel = fsdp_param.sharded_size.numel()
new_sharded_grad = reduce_scatter_output[
    flat_grad_offset : flat_grad_offset + sharded_numel
].view(fsdp_param.sharded_size)
```
to
```
new_sharded_grad = torch.as_strided(
    reduce_scatter_output,
    size=fsdp_param.sharded_size,
    stride=fsdp_param.sharded_stride,
    storage_offset=flat_grad_offset,
)
```
as affecting correctness for a 2D unit test downstream. The strange thing is that I checked that the two versions pass `torch.equal()` and have the same size/stride/data pointer, so I am not sure where the difference is coming from.

cc: @yifuwang

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:173]: I have further found that (1) it seems to only affect the 2D-parallel case and (2) it seems to only affect uneven case (when dim-0 is not divisible by world size).

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:173]: It seems possible that the `DTensor`'s `_local_tensor` stride is not correct for uneven case 🤔 

I am seeing some weirdness that needs to be investigated more. Like the uneven chunk from `torch.chunk()` shows stride `(5, 1)` despite shape `(1, 4)` (so we expect stride `(4, 1)`).

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:173]: I figured it out: the `fsdp_param.sharded_stride` was being taken from the `sharded_param` chunked from the `DTensor._local_tensor`, which had the incorrect stride in the uneven sharding case. Therefore, the `fsdp_param.sharded_stride` being passed to `torch.as_strided` was wrong. I am not sure why the direct `.strides()` check did not fail when comparing the two approaches -- I must have done something wrong.

I filed an issue and added a workaround in this PR.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 7]
```diff
@@ -63,6 +63,27 @@ def _init_mp_dtypes(self) -> None:
         self._orig_dtype = next(iter(orig_dtypes))
         self._param_dtype = self._orig_dtype
 
+    def _init_grad_divide_factors(self):
+        """
+        For N data parallel workers, each worker computes g_i, and they
+        collectively reduce to compute (g_1 + ... + g_N) / N. To avoid overflow
+        and underflow, we divide by sqrt(N) before and after the reduction.
+        """
+        data_parallel_world_size = 1
```
Comment (wconstab) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:72]: why do you init this to 1 and multiply out?

also the comment says sqrt(N) but aren't you doing something that is loosely inspired by that but not equivalent?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py (hunk 6 above)
Comment (yifuwang) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:173]: >  figured it out: the fsdp_param.sharded_stride was being taken from the sharded_param chunked from the DTensor._local_tensor

Oh yea I noticed that when I first saw it. The PR didn't have usage for it yet so I wasn't sure if it was intended that way.

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 8]
```diff
@@ -178,6 +179,10 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
         chunks = _chunk_with_empty(param_data, shard_world_size, dim=0)
         sharded_param = chunks[shard_rank]
         self.sharded_size = sharded_param.size()
+        # TODO: DTensor's local tensor may not have the correct stride for the
+        # uneven sharding case, so we construct a new tensor to get the stride:
+        # https://github.com/pytorch/pytorch/issues/118836
+        self.sharded_stride = torch.empty(self.sharded_size).stride()
```
Comment (wconstab) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:185]: we just want to compute the contiguous strides of a particular shape?  cant we do it less cheaply than allocating a tensor- perhaps we have helpers for this but only in c++ and no bindings? probably fine to leave like this.  (only happens on init path so nbd i guess)

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 9]
```diff
@@ -105,6 +110,118 @@ def foreach_all_gather_copy_out(
...
+    unsharded_grads.clear()
+    # The RS output is allocated in the RS stream and used in the default
+    # stream (for optimizer). To ensure its memory is not reused for later
+    # RSs, we do not need extra synchronization since the sharded parameters
+    # hold refs through the end of backward.
+    return reduce_scatter_view_out_event
+
+
+def foreach_reduce_scatter_copy_in(
+    fsdp_params: List[FSDPParam],
```
Comment (wconstab) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:198]: why is fsdp_params an arg here, unused?

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 8 above)
Comment (yifuwang) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:185]: fwiw there's this:

```python
>>> torch._prims_common.make_contiguous_strides_for(torch.Size([20, 10, 5]))
(50, 5, 1)
```

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 7 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:72]: Looking at https://github.com/facebookresearch/fairscale/pull/565, I think this 'init to 1 and multiply out' is just a nice way of making sure the pre- and post-factors multiply to the world size while keeping them as integer values. I did not see any motivation for why they need to be integer though.

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 8 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py]: Good suggestions! For clarity, I renamed the variable to `self.contiguous_sharded_stride` and used `make_contiguous_strides_for()` (still at initialization time).
